torch/_dynamo/__init__.py

```diff
@@ -156,14 +162,96 @@ def mark_dynamic(t, index):
     4) Attempts to trace this function will explicitly raise. As such, all calls to mark_dynamic must be made
     before torch.compile.
 
+    """
+    mark_dynamic_constrain(t, index, min=None, max=None)
+
+
+@forbid_in_graph
+def mark_dynamic_constrain(
+    t, index, *, min: Optional[int] = None, max: Optional[int] = None
+):
+    """
+    To fully understand this API, please read [Note - on the state of mark_dynamic] first,
+    as this API is an enrichment over that API.
+
+    In its current state, mark_dynamic_constrain fully subsumes mark_dynamic.
+
+    mark_dynamic_constrain allows users to provide a directive that the dimension will fall
+    within a given range. A range can be unbounded on either min, or max. Multiple calls to this API for
+    the same dimension will take the most conservative intersection of all ranges. At guard accumulation
```

Comment:
Alternately, you could just error.

There is a little bit of an open UX question here, but erroring leaves the door open for either semantics.

Reply:
Because the marker is EXTERNAL to the model, it is not obvious that intersection is better than overwriting. In particular, overwriting is more expressive than intersection.

Reply:
No strong opinion! Can swap to overwriting.

Reply:
discussed offline, swapped to first or clear.